### PR TITLE
Feature/cache

### PR DESCRIPTION
**Pull request by Jibbarth**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Fixed tickets | #... 

:rocket: This PR add cache system to avoid recheck a file if it's content doesn't change between two analysis  

See the [Blackfire comparaison profile](https://blackfire.io/profiles/compare/486f8215-f0f1-43d9-b649-11abf22bcbd3/graph) between 2 analysis.

See demo: 
phpinsights-cache

TODO: 
- [x] Add parameter to force full analysis without cache 
- [x] Handle invalidate cache if a file is added to excluded file ?

I think this feature will allow us after to try implement parallelization (#390), as we'll can retrieve issues from cache, so we can split the analysis between multiple process.

### CODE REVIEW

File: src/Application/Injectors/Cache.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+{
+    /**
+     * Inject Cache into the container definitions.
+     *
+     * @return array<string, callable>
+     */
+    public function __invoke(): array
+    {
+        return [
+            CacheInterface::class => static fn (): CacheInterface => new Psr16Cache(
```
Comment (nunomaduro) [inline src/Application/Injectors/Cache.php:24]: PHP 7.4. <3

File: tests/TestCase.php [hunk 2]
```diff
@@ -93,6 +96,10 @@ abstract class TestCase extends BaseTestCase
         $configurationDefinition = $container->extend(Configuration::class);
         $configurationDefinition->setConcrete($config);
 
+        // disable cache
+        $cacheDefinition = $container->extend(CacheInterface::class);
+        $cacheDefinition->setConcrete(new Psr16Cache(new FilesystemAdapter('phpinsight', 1, '/dev/null')));
```
Comment (nunomaduro) [inline tests/TestCase.php:101]: I would expect something like `sys_get_temp_dir`. I am missing something?

**Comment by olivernybroe** [comment]
Nice! The implementation is really clean.

Out of curiosity, did you look into adding the cache globally, so we won't have to write cache for each runner?

File: tests/TestCase.php (hunk 2 above)
Comment (Jibbarth) [reply tests/TestCase.php]: In fact, two tests was launched on same directory for test exclude configuration. 
I remove this modification by adding a configuration cache key :+1:

**Comment by Jibbarth** [comment]
> Out of curiosity, did you look into adding the cache globally, so we won't have to write cache for each runner?

:thinking: it must be doable, but I thought it would allow us to adding parallelization more easily by putting cache by runners.

**Comment by olivernybroe** [comment]
@Jibbarth Funny that you mention parallelization, because my thought were the exact same, just that it must be easier if the caching was before the runners. 

If the cache is up to date, then we never have to touch any runners :) 
But I think we really have to think about this one, as it sounds like you might have some insights here that I don't fully follow. 

Some of the points I see with a centralized cache is 
- No need to write cache for each runner (this makes the runners simpler to create)
- The data of the cache would be easier to understand, as it is not tied to a runner
- Less processing of the data, as if we cache the result from the runner, we would have to their result to our format before converting it to the desired output, whereas the centralized cache can cache it in our format.
- Less flexibility (as centralized cache has no idea how the runner works, it cannot touch any specific stuff related to phpcs, if this would be needed)

File: src/Domain/Contracts/DetailsCarrier.php [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+declare(strict_types=1);
+
+namespace NunoMaduro\PhpInsights\Domain\Contracts;
+
+use NunoMaduro\PhpInsights\Domain\Details;
+
+/**
+ * @internal
+ */
+interface DetailsCarrier extends HasDetails
```
Comment (Jibbarth) [inline src/Domain/Contracts/DetailsCarrier.php:12]: Not sure about naming here, if someone has a better idea :upside_down_face:

**Comment by Jibbarth** [comment]
@olivernybroe I took in consideration your advices. 

I think you are right, it's simplier and file processors don't have to handle them. 

For parallelization, as I think we could split file list in smaller batch, and for each, call runner on it, that should work also. :+1:

File: src/Domain/Contracts/DetailsCarrier.php (hunk 3 above)
Comment (olivernybroe) [reply src/Domain/Contracts/DetailsCarrier.php:12]: Could you explain the purpose if the interface? 😄

File: src/Application/Console/Commands/AnalyseCommand.php [hunk 4]
```diff
@@ -55,6 +57,11 @@ public function __invoke(InputInterface $input, OutputInterface $output): int
 
         $formatter = FormatResolver::resolve($input, $output, $consoleOutput);
 
+        // flush cache before processing if asked
+        if ($input->getOption('flush-cache') === true) {
```
Comment (olivernybroe) [inline src/Application/Console/Commands/AnalyseCommand.php:61]: You can use `hasOption` her instead?

Comment (Jibbarth) [reply src/Application/Console/Commands/AnalyseCommand.php:61]: No, `hasOption` check if there is option in definition, not if option is setted 😕

**Comment by Jibbarth** [comment]
@olivernybroe 

> All of the output is showed from data by the Details object anyway

It's not true, we display the title and the InsightClass from Insights Objects. That's why I think a CacheInsight class couldn't work :slightly_frowning_face: 

For `DetailCarrier` interface, the purpose is to mean : this class transports probably lots of details, and details can be added over the time

**Comment by Jibbarth** [comment]
For information, there is the last Blackfire performance comparison between Master and feature/cache :wink:  

https://blackfire.io/profiles/compare/aad3ad58-25a0-45ca-b895-95f3b9340f31/graph

:rocket: Gain 86% on performance :muscle:

**Comment by olivernybroe** [comment]
@Jibbarth that's an amazing improvement with cache, love it! 

Hmm what do you say if we change it so Details object actually has all of the data instead? I think that makes sense and it simplifies the cache and future stuff. 

It might also help us simplify the formatters, as they only have to look at detail objects.

It can also just be something for a pr later to not expland this pr too much

**Comment by Jibbarth** [comment]
@olivernybroe Thanks for the review :+1: 

> Hmm what do you say if we change it so Details object actually has all of the data instead? Or would that be weird?

I don't see the need for now, but maybe. Anyway that would be a huge refacto.

**Pull request merged**